src/engine/sparqlExpressions/AggregateExpression.h

```diff
@@ -203,40 +262,37 @@ class CountExpression : public CountExpressionBase {
       return std::nullopt;
     }
   }
+  ValueId resultForEmptyGroup() const override { return Id::makeFromInt(0); }
 };
 
-// Take a `NumericOperation` that takes numeric arguments (integral or floating
-// points) and returns a numeric result. Return a function that performs the
-// same operation, but takes and returns the `NumericValue` variant.
-template <typename NumericOperation>
-inline auto makeNumericExpressionForAggregate() {
-  return [](const std::same_as<NumericValue> auto&... args) -> NumericValue {
-    auto visitor = []<typename... Ts>(const Ts&... t) -> NumericValue {
-      if constexpr ((... || std::is_same_v<NotNumeric, Ts>)) {
-        return NotNumeric{};
-      } else {
-        return (NumericOperation{}(t...));
-      }
-    };
-    return std::visit(visitor, args...);
-  };
-}
-
-// SUM
+// Aggregate expression for SUM.
 inline auto addForSum = makeNumericExpressionForAggregate<std::plus<>>();
-using SumExpression = AGG_EXP<decltype(addForSum), NumericValueGetter>;
+using SumExpressionBase = AGG_EXP<decltype(addForSum), NumericValueGetter>;
+class SumExpression : public AGG_EXP<decltype(addForSum), NumericValueGetter> {
+  using SumExpressionBase::SumExpressionBase;
+  ValueId resultForEmptyGroup() const override { return Id::makeFromInt(0); }
```

Comment:
I think you can make the defaultResultForEmptyGroup a template parameter of the `AggregateExpression` class,
Then all this code goes away and you can again write something like 
```
using SumExpression = AGG_EXP<decltype(addForSum), NumericValueGetter, Id::makeFromInt(0)>;
```
(Finally a good use of all the `makeFrom...` functions being constexpr (if I'v e forgotten one of them, please add the constexpr).

Reply:
My first implementation used a template parameter (slightly differently from what you suggest, but in principle the same). But I didn't like the additional code required in the `.cpp` file: the `template ...` in front of all the function definitions becomes longer then and so do the explicit instantiations int the end.

Do you have a strong argument in favor of the template parameter vs. an additional member function? Efficiency is not an issue here as far as I can see (because it's only for a very special case with a single group).

Reply:
I understand your reasoning here, we definitely don't want the templates to become even harder to read.
I have thought about it, and I think I have found a way to hide this
additional argument in the `Operation` class, s.t. you only have to add the single value. in a single place.

Reply:
I am leaving this discussion open until I have implemented this.